lib/Sema/CSSimplify.cpp

```diff
@@ -2218,9 +2217,11 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
   };
 
   auto implodeParams = [&](SmallVectorImpl<AnyFunctionType::Param> &params) {
+    // Form an imploded tuple type, dropping the parameter flags as
+    // canImplodeParams makes sure we're not dealing with vargs, inout, etc,
+    // but we may still have e.g ownership flags left over, which aren't useful.
     auto input = AnyFunctionType::composeTuple(getASTContext(), params,
-                                               /*canonicalVararg=*/false);
-
+                                               /*wantParamFlags*/ false);
```

Comment:
Wouldn't canImplodeParams() return false in this case?

Reply:
It only currently returns false for vargs, inout, and autoclosure, but there may still be other parameter flags like `__owned` or `@_nonEphemeral` to strip.

Reply:
I guess the question is whether `canImplodeParams` should return false if any of the parameters here have flags? I think the answer is yes if it's only valid to convert "simple" parameters lists into tuples otherwise there is a risk that some new parameter flag would produce a valid solution but then crash in runtime?

Reply:
That's a fair point, I *believe* we should support stripping for ownership flags such as `__owned`, as we can handle the function conversion through a thunk, and we have test cases for that in the test suite. I believe we should also support stripping for `@_nonEphemeral`, as the attribute can be freely stripped off a function type anyways. Talking with @rxwei, it seems we should also support stripping `@noDerivative` (as long as the destination isn't `@differentiable`).

However I do agree that defaulting to returning false from `canImplodeParams` for any other parameter flag makes sense. For `isolated` in particular that seems like the right behavior.

I've gone ahead and updated the PR to default to returning false, but with some exceptions for the above mentioned cases.